app/src/main/java/com/codesoom/assignment/application/ProductService.java

```diff
@@ -0,0 +1,92 @@
+package com.codesoom.assignment.application;
+
+import com.codesoom.assignment.domain.Product;
+import com.codesoom.assignment.domain.ProductRepository;
+import com.codesoom.assignment.errors.ProductNotFoundException;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.List;
+
+/**
+ * 상품을 관리합니다.
+ */
+@Service
+@Transactional
+public class ProductService {
+
+    private final ProductRepository productRepository;
+
+    /**
+     * 주어진 의존성 객체를 사용하여 ProductService 객체를 생성합니다.
+     * @param productRepository 상품 repository
+     */
+    public ProductService(ProductRepository productRepository) {
+        this.productRepository = productRepository;
+    }
+
+    /**
+     * 저장하고 있는 상품 목록을 반환합니다.
+     * @return 상품 목록
+     */
+    public List<Product> getProducts() {
+        return productRepository.findAll();
+    }
+
+    /**
+     * 주어진 id와 일치하는 상품을 반환합니다.
+     * @param id 상품 id
+     * @return 일치하는 id의 상품
+     */
+    public Product getProduct(Long id) {
+        return productRepository.findById(id)
+                .orElseThrow(() -> new ProductNotFoundException(id));
+    }
+
+    /**
+     * 상품을 저장하고 저장된 상품을 반환합니다.
+     * @param source 저장할 상품의 source
+     * @return 저장된 상품
+     */
+    public Product saveProduct(Product source) {
+        Product.createSaveProduct(
+                source.getName(),
+                source.getMaker(),
+                source.getPrice(),
+                source.getImageUrl()
+        );
+
+        return productRepository.save(source);
+    }
+
+    /**
+     * 주어진 id의 상품을 수정하고 수정된 상품을 반환합니다.
+     * @param id 수정할 상품 id
+     * @param source 수정할 상품 source
+     * @return 수정된 상품
+     */
+    public Product updateProduct(Long id, Product source) {
+        Product product = productRepository.findById(id)
+                .orElseThrow(() -> new ProductNotFoundException(id));
+
+        product.update(
+                source.getName(),
+                source.getMaker(),
+                source.getPrice(),
+                source.getImageUrl()
+        );
+
+        return product;
+    }
+
+    /**
+     * 주어진 id의 상품을 삭제합니다.
+     * @param id 삭제할 상품 id
+     */
+    public void deleteProduct(Long id) {
+        Product product = productRepository.findById(id)
+                .orElseThrow(() -> new ProductNotFoundException(id));
```

Comment:
왜 예외를 던졌고, 그로 인해 어떤 일이 좌절되었는지를 설명해줘야 디버깅하는 사람이 야근을 덜 합니다. 로그 파일에 숫자 하나만 덜렁 있으면 문제를 파악할 때 좀 더 어렵겠죠.

```suggestion
                .orElseThrow(() -> new ProductNotFoundException(String.format("아이디 [%d]에 해당하는 상품을 찾지 못했으므로, 상품을 삭제하지 못했습니다.", id)));
```

Reply:
네 맞아요. 로그 파일만 봐도 원인을 알 수 있는 게 좋은 코드죠!